### ISSUE TITLE
onProgress, totalSize=1

### ISSUE THREAD

**Issue by LondonJammo** [loopj/android-async-http#957]
Hi, I'm using 1.4.8 and when downloading a file, in the onProgress() event, totalSize is always=1. Is the filesize not known prior to download? Should I get the filesize another way before starting the download? The reason I ask is because I'm trying to implement a ProgressDialogBar.

Lib:
`compile 'com.loopj.android:android-async-http:1.4.8'`

Code:
```Java
 client.get(url, new AsyncHttpResponseHandler() {

            @Override
            public void onProgress(long bytesWritten, long totalSize) {
                super.onProgress(bytesWritten, totalSize);

                Log.v("myapp","DOWNLOAD PROGRESS -> Written:" + bytesWritten + " Total:" + totalSize);
```

Log:
`V/myapp﹕ DOWNLOAD PROGRESS -> Written:228913 Total:1`

**Comment by smarek** [comment]
Looks like you're dealing with compressed output, when the response is Gzip (or other) compressed, the size can be unknown. However in 1.4.8 this should be fixed since #647 resolution
Do you really use latest version of 1.4.8 ? Could you please test your use-case with 1.4.9-SNAPSHOT?
Thanks you !

**Comment by LondonJammo** [comment]
`Error:(33, 13) Failed to resolve: com.loopj.android:android-async-http:1.4.9-SNAPSHOT`

**Comment by smarek** [comment]
please see readme for snapshot repository url

**Comment by LondonJammo** [comment]
Yep, same behaviour with the snapshot, still returning value 1

**Comment by smarek** [comment]
Can you please check response headers, if there is anything about response size? Because it's possible that in this case the final size could be really unknown.

**Comment by LondonJammo** [comment]
the `onSuccess(int statusCode, Header[] headers, byte[] response)` returns the correct filesize. It's in `BufferedHeader` with `Content-Length: 5508` (expected value). However, the onSuccess only fires when the full data has been downloaded, and not during onProgress, which is what I need. OnProgress doesn't offer any headers to check for filesize

**Comment by LondonJammo** [comment]
Any update on this? I'm wondering whether I need to change to an alternative library/method unless this issue is planning on being fixed soon. Many thanks

**Comment by smarek** [comment]
Hey, I'm sorry but now it's priority to finish 1.4.9.1 integration problems.

However I have probably solution for you.
Implement method `void onPreProcessResponse(ResponseHandlerInterface instance, HttpResponse response)` in request callback (ResponseHandlerInterface subclass)

There you can grab the headers and base the progress on this.

**Comment by LondonJammo** [comment]
Ok thanks. Had a look, and added an `@Override`. Stepped through the available data but there doesn't seem to be anything in `instance` or `response` with the response size.

```java
client.get(this.Url, new AsyncHttpResponseHandler() {
            @Override
            public void onPreProcessResponse(ResponseHandlerInterface instance, HttpResponse response) {
                super.onPreProcessResponse(instance, response);
            }
   ...
```

**Comment by smarek** [comment]
Previously you claimed there are headers in response that indicate final data size, so I started with that idea.
Only thing could be that this header is added only after the request is full consumed, could you please verify this?

**Comment by LondonJammo** [comment]
Yes that's right, but only in the `onSuccess(int statusCode, Header[] headers, byte[] response)` override. However, that only runs when the response has been completely downloaded. I can't see anything in the `onPreProcessResponse()` Override that shows filesize. Am I looking in the wrong place?

**Comment by smarek** [comment]
That could mean the header `Content-Length` is added only after the content is consumed, which seems strange to me.
Could you please provide example URL where I could test the scenario with current version of library, maybe providing more help to you?

**Comment by LondonJammo** [comment]
Ok, I think I need to eat some humble pie. I found the Content-Length header in the `onPreProcessResponse` under the `response` variable. Apologies. My code for retreiving is is below, and I set it to a local variable, which is then used `onProgress(){}`

```java
            @Override
            public void onPreProcessResponse(ResponseHandlerInterface instance, HttpResponse response) {
                super.onPreProcessResponse(instance, response);
                Header[] headers = response.getAllHeaders();
                for (Header header : headers) {
                    if (header.getName().equalsIgnoreCase("content-length")) {
                        String value = header.getValue();
                        int valueSize = Integer.valueOf(value);
                        setFileSizeFromHeader(valueSize);
                    }
                }
            } 

            @Override
            public void onProgress(long bytesWritten, long totalSize) {
                super.onProgress(bytesWritten, totalSize);

                //int progress = Math.round((bytesWritten/totalSize)*100);
                int progress = Math.round((bytesWritten/getFileContentSize())*100);
                Log.v("***", "DOWNLOAD PROGRESS -> Written:" + bytesWritten + " Total:" + getFileContentSize() + " Progress:" + progress + "%");

            }
```

*However*, I've noticed that with smaller sizes (eg requesting the http response www.example.com), the Content-Length is correct. When I try and access a JSON object on my server, the response size is 42911, but the `Content-Length` header only says 12374. Are multiple requests being done?

My OnProgress log ends with:
```java
...
09-25 21:48:00.682  28905-28905/com.*** V/***﹕ DOWNLOAD PROGRESS -> Written:34988 Total:12374 Progress:200%
09-25 21:48:00.682  28905-28905/com.*** V/AsyncHttpRH﹕ Progress 36794 from 12374 (297%)
09-25 21:48:00.682  28905-28905/com.*** V/***﹕ DOWNLOAD PROGRESS -> Written:36794 Total:12374 Progress:200%
09-25 21:48:00.682  28905-28905/com.*** V/AsyncHttpRH﹕ Progress 38533 from 12374 (311%)
09-25 21:48:00.682  28905-28905/com.*** V/***﹕ DOWNLOAD PROGRESS -> Written:38533 Total:12374 Progress:300%
09-25 21:48:00.683  28905-28905/com.*** V/AsyncHttpRH﹕ Progress 40383 from 12374 (326%)
09-25 21:48:00.683  28905-28905/com.*** V/***﹕ DOWNLOAD PROGRESS -> Written:40383 Total:12374 Progress:300%
09-25 21:48:00.683  28905-28905/com.*** V/AsyncHttpRH﹕ Progress 42075 from 12374 (340%)
09-25 21:48:00.684  28905-28905/com.*** V/***﹕ DOWNLOAD PROGRESS -> Written:42075 Total:12374 Progress:300%
09-25 21:48:00.684  28905-28905/com.*** V/AsyncHttpRH﹕ Progress 42911 from 12374 (347%)
09-25 21:48:00.684  28905-28905/com.*** V/***﹕ DOWNLOAD PROGRESS -> Written:42911 Total:12374 Progress:300%
```

**Comment by smarek** [comment]
Cool to know my theory was good, just an idea, you don't use any static or shared field to store `Content-Length` contents in `setFileSizeFromHeader()` ?
The length indicates size of compressed content (number of bytes you should read from server before the response is read out successfully)

Two things:
1) Content-Length is stored and operated within library as Long, so maybe you could lose precision with large files (see #847)
2) Are you extending directly `AsyncHttpResponseHandler` or using any specific subclass (ie. FileAsyncHttpResponseHandler ) ?

**Comment by vagrant1991** [comment]
the same problem, @smarek  how can i solve it?

11-24 11:34:07.973 29453-29519/com.seeyon.cmp V/AsyncHttpRH: Progress 1088 from 2007 (54%)
11-24 11:34:08.003 29453-29519/com.seeyon.cmp V/AsyncHttpRH: Progress 2234 from 2007 (111%)
11-24 11:34:08.003 29453-29519/com.seeyon.cmp V/AsyncHttpRH: Progress 4398 from 2007 (219%)
11-24 11:34:08.003 29453-29519/com.seeyon.cmp V/AsyncHttpRH: Progress 5442 from 2007 (271%)

**Comment by smarek** [comment]
@vagrant1991 you can debug the library and try to explain the issue, or you can directly contribute by providing pull-request

**Comment by vagrant1991** [comment]
@smarek  the totalSize is get from content-length, but sometimes there is no content-length in response header or the content-length is incorrect.

**Comment by vagrant1991** [comment]
hi ,everybody, so the real question is how to get the correct totalSize if there is no content-length in response header or the content-length is incorrect.
